### PR TITLE
Restarting OpenMP work

### PR DESCRIPTION
**Pull request by Shillaker**
Here I wanted to remove some of the complexity in the existing OpenMP implementation to make upcoming modifications more straightforward. Changes:

- Removed local thread pool
- Tidied up some of the `omp` test functions
- Merged all other OpenMP-related code into a single header and .cpp file (as there's not _too_ much of it left)
- Slightly rearranged the OpenMP host interface file to group similar functions together
- Formatting

Side note: the release build is failing but I'm getting on logs from Github Actions. The toolchain release build was also failing, not sure if they're connected. I've run the release builds on our machines and they worked fine. I don't have time to try and fix it now so I propose we merge this and continue.

### CODE REVIEW

File: docs/openmp.md [hunk 1]
```diff
@@ -20,21 +20,3 @@ inv upload omp hellomp
...
-
-To add a new function you need to include it in both the `libfaasmp` definition,
-and the Faasm host interface.
-
-LLVM adds GNU compatibility data symbols on some features for synchronisation.
-The wasm linker needs to be aware of the sizes and the names of the symbols,
-which are not valid C/C++ identifiers, and as such are resolved in assembly in
-`libfaasmp/z_Linux_asm.S` in the `faasm-toolchain` repo. An error from the
-linker when compiling new parts of the API implies you need to you need to add
-the symbol to this file.
```
Comment (Shillaker) [inline docs/openmp.md:40]: Moved to faasm-toolchain repo with libopenmp

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,7 +6,6 @@ option(FAASM_STATIC_LIBS "Statically link Faasm libs" ON)
 
 # Performance functionality
 option(FAASM_SELF_TRACING "Turn on system tracing using the logger" OFF)
-option(FAASM_OPENMP_TRACING "Trace OpenMP forks in Redis" OFF)
```
Comment (Shillaker) [inline CMakeLists.txt:9]: We'll be changing the implementation so won't need this

File: include/wavm/WAVMWasmModule.h [hunk 3]
```diff
@@ -129,8 +127,6 @@ class WAVMWasmModule final
 
     uint32_t allocateThreadStack();
 
-    std::unique_ptr<openmp::PlatformThreadPool>& getOMPPool();
```
Comment (Shillaker) [inline include/wavm/WAVMWasmModule.h:132]: The thread pool was good for local performance, but added complexity which we need to remove in order to redo the implementation

File: include/wavm/ThreadState.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <mutex>
```
Comment (Shillaker) [inline include/wavm/ThreadState.h:3]: I've merged what's left from the deleted headers in the rest of this diff into this single file to make it a bit clearer what's going on

File: src/wavm/ThreadState.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
+#include <wavm/ThreadState.h>
+#include <wavm/WAVMWasmModule.h>
```
Comment (Shillaker) [inline src/wavm/ThreadState.cpp:2]: As with the `ThreadState.h` file, this is a single place for all the implementation parts we'll be keeping

File: src/wavm/WAVMWasmModule.cpp [hunk 6]
```diff
@@ -1564,31 +1559,6 @@ Runtime::Function* WAVMWasmModule::getFunctionFromPtr(int funcPtr)
...
-          std::make_shared<openmp::SingleHostLevel>());
-    }
-
-    openmp::setTLS(msg.ompthreadnum(), ompLevel);
-}
-
-std::unique_ptr<openmp::PlatformThreadPool>& WAVMWasmModule::getOMPPool()
-{
-    return OMPPool;
-}
```
Comment (Shillaker) [inline src/wavm/WAVMWasmModule.cpp:1590]: Preparation of the OpenMP context is now defined in the OpenMP-specific files.

File: src/wavm/openmp.cpp [hunk 7]
```diff
@@ -1,22 +1,27 @@
-#include "wavm/openmp/openmp.h"
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:1]: This file contains the OpenMP host interface functions, so is the single entrypoint to all the OpenMP code

File: src/wavm/openmp.cpp [hunk 8]
```diff
@@ -1,22 +1,27 @@
...
-using namespace openmp;
+
+struct LocalThreadArgs
+{
+    int tid = 0;
+    std::shared_ptr<Level> level = nullptr;
+    WAVMWasmModule* parentModule;
+    faabric::Message* parentCall;
+    WasmThreadSpec spec;
+};
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:24]: This used to live in a header but is only used inside this file now

File: src/wavm/openmp.cpp [hunk 9]
```diff
@@ -66,7 +72,7 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
                                omp_get_thread_num)
 {
     faabric::util::getLogger()->debug("S - omp_get_thread_num");
-    return thisThreadNumber;
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:69]: The OpenMP context was previously held in a number of global thread-local variables, so I've now encapsulated this into an `OpenMPContext` class accessible via the `getOpenMPContext()` method

File: src/wavm/openmp.cpp [hunk 10]
```diff
@@ -1,60 +1,22 @@
...
- * Performs actual static assignment
- */
-template<typename T>
-void for_static_init(I32 schedule,
-                     I32* lastIter,
-                     T* lower,
-                     T* upper,
-                     T* stride,
-                     T incr,
-                     T chunk);
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:31]: I've grouped all the `for_static_init` functions together later in the file

File: src/wavm/openmp.cpp [hunk 11]
```diff
@@ -1,60 +1,22 @@
...
 
-/**
- * Function used to spawn OMP threads. Will be called from within a thread
- * (hence needs to set up its own TLS)
- */
-I64 ompThreadEntryFunc(void* threadArgsPtr)
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:37]: This function is no longer needed

File: src/wavm/openmp.cpp [hunk 12]
```diff
@@ -118,14 +80,67 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
...
         logger->warn("Trying to set active level with a negative number {}",
                      level);
-        return;
+    } else {
+        getOpenMPContext().level->maxActiveLevels = level;
     }
-    thisLevel->maxActiveLevel = level;
 }
 
+WAVM_DEFINE_INTRINSIC_FUNCTION(env,
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:92]: These functions are unchanged, I've just grouped those related to number of threads together nearer the top.

File: src/wavm/openmp.cpp [hunk 13]
```diff
@@ -473,73 +470,81 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
         }
 
         logger->debug("Distributed Fork finished successfully");
-    } else { // Single host
+
+    } else if (conf.threadMode == "local") {
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:474]: This block has changed the most in this file. It used to use the thread pool and hence introduced a couple of layers of indirection. Now it just spawns threads using `std::thread`, so should be more self-contained. This is important as this section is likely to change a lot when we change the implementation.

File: src/wavm/openmp.cpp [hunk 14]
```diff
@@ -699,35 +836,41 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
...
-        case ReduceTypes::emptyBlock:
-            retVal = 1;
-            break;
-        case ReduceTypes::atomicBlock:
-            retVal = 2;
-            break;
-        case ReduceTypes::notDefined:
-            throw std::runtime_error("Unsupported reduce operation");
-            break;
-        case ReduceTypes::multiHostSum:
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:726]: These `ReduceTypes` appeared not to be used very much, so I've removed them.

File: src/wavm/openmp.cpp [hunk 15]
```diff
@@ -766,22 +909,22 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
...
                                I32 loc,
                                I32 gtid,
-                               I32 num_vars,
-                               I32 reduce_size,
-                               I32 reduce_data,
-                               I32 reduce_func,
+                               I32 numVars,
+                               I32 reduceSize,
+                               I32 reduceData,
+                               I32 reduceFunc,
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:915]: CamelCase

File: func/omp/mt_pi.cpp [hunk 16]
```diff
@@ -2,42 +2,42 @@
...
                 result++;
             }
         }
     }
 
-    double pi = (4.0 * result) / iterations;
+    double pi = (4.0 * result) / ITERATIONS;
 
-    if (pi - 3.14 > 0.01) {
+    if (abs(pi - 3.14) > 0.01) {
```
Comment (Shillaker) [inline func/omp/mt_pi.cpp:36]: This wasn't checking if `pi` was much smaller than 3.14 before

File: func/omp/mt_pi.cpp [hunk 17]
```diff
@@ -2,42 +2,42 @@
...
 int main(int argc, char** argv)
 {
-    long long iterations = 1000LL;
-    int num_threads = 1;
-    if (argc == 3) {
-        num_threads = std::stoi(argv[1]);
-        iterations = std::stoll(argv[2]);
-    } else if (argc != 1) {
-        printf("Usage: mt_pi [num_threads num_iterations]");
-    }
```
Comment (Shillaker) [inline func/omp/mt_pi.cpp:20]: I couldn't find any use for this being configurable

File: func/omp/nested_levels_test.cpp [hunk 18]
```diff
@@ -1,236 +1,191 @@
 /**
- *  Checks a few invariants of nested levels.
```
Comment (Shillaker) [inline func/omp/nested_levels_test.cpp:2]: This was really complicated and hard to follow, plus it didn't actually pass even on native openmp for me, so I've cut it down to be simpler

File: func/omp/setting_num_threads.cpp [hunk 19]
```diff
@@ -1,20 +1,24 @@
 #include <cstdio>
 #include <omp.h>
 
-int get_expected_value(int total)
```
Comment (Shillaker) [inline func/omp/setting_num_threads.cpp:4]: Refactored this one a bit too

File: func/omp/simple_reduce.cpp [hunk 20]
```diff
@@ -1,19 +1,29 @@
 #include <cstdio>
-#include <faasm/faasm.h>
 #include <omp.h>
```
Comment (Shillaker) [inline func/omp/simple_reduce.cpp:3]: Refactored this to be a bit more readable

File: func/omp/stack_debug.cpp [hunk 21]
```diff
@@ -1,58 +0,0 @@
...
-#include <faasmp/faasmp.h>
-#include <omp.h>
-#include <stdio.h>
-
-/**
- * This function checks that OMP threads each have their own private stack.
- * The inner loop will always yield the same wasm offsets to refer to the
- * stack variables (as the code is exactly the same), but the runtime must
- * map these onto _separate_ stacks behind the scenes, rather than map them
- * all onto the same memory locations.
```
Comment (Shillaker) [inline func/omp/stack_debug.cpp:11]: This relies on the `__faasm_debug_copy` operation which I've removed to simplify things

File: src/wasm/CMakeLists.txt [hunk 22]
```diff
@@ -17,4 +17,4 @@ set(LIB_FILES
         )
 
 faasm_private_lib(wasm "${LIB_FILES}")
-target_link_libraries(wasm storage openmp)
+target_link_libraries(wasm storage)
```
Comment (Shillaker) [inline src/wasm/CMakeLists.txt:20]: No longer a separate openmp library

File: src/wavm/WAVMWasmModule.cpp [hunk 23]
```diff
@@ -989,7 +994,11 @@ U32 WAVMWasmModule::mmapFile(U32 fd, U32 length)
 
 U32 WAVMWasmModule::allocateThreadStack()
 {
-    return mmapMemory(THREAD_STACK_SIZE);
+    createMemoryGuardRegion();
+    U32 wasmPtr = mmapMemory(THREAD_STACK_SIZE);
+    createMemoryGuardRegion();
```
Comment (Shillaker) [inline src/wavm/WAVMWasmModule.cpp:999]: Stacks that we provision ourselves should always be sandwiched between guard regions to catch over/ underflows

File: src/wavm/dynlink.cpp [hunk 24]
```diff
@@ -73,7 +73,7 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env, "dlerror", I32, dlerror)
     faabric::util::getLogger()->debug("S - dlerror");
     std::string errorMessage("Wasm dynamic linking error. See logs");
 
-    wasm::WAVMWasmModule* module = getExecutingWAVMModule();
+    WAVMWasmModule* module = getExecutingWAVMModule();
```
Comment (Shillaker) [inline src/wavm/dynlink.cpp:76]: There seemed to be some unnecessary namespacing that started erroring when I changed the layout, so I removed it

File: src/wavm/openmp.cpp [hunk 25]
```diff
@@ -1,60 +1,37 @@
...
-    // Set up various TLS
-    setTLS(args.tid, args.level);
-    setExecutingModule(args.parentModule);
-    setExecutingCall(args.parentCall);
-    return getExecutingWAVMModule()->executeThreadLocally(args.spec);
-}
+#define OMP_FUNC(str)                                                          \
+    auto ctx = getOpenMPContext();                                             \
+    auto logger = faabric::util::getLogger();                                  \
+    logger->debug("OMP {}: " str, ctx.threadNumber);
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:25]: Added some convenience macros

File: src/wavm/openmp.cpp [hunk 26]
```diff
@@ -452,147 +456,242 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
...
-    const std::shared_ptr<spdlog::logger>& logger = faabric::util::getLogger();
-    logger->debug("S - __faasmp_debug_copy {} {}", src, dest);
+    if (ctx.level->numThreads == 1) {
+        *lastIter = true;
 
-    // Get pointers on host to both src and dest
-    Runtime::Memory* memoryPtr = getExecutingWAVMModule()->defaultMemory;
-    int* hostSrc = &Runtime::memoryRef<int>(memoryPtr, src);
-    int* hostDest = &Runtime::memoryRef<int>(memoryPtr, dest);
+        if (incr > 0) {
```
Comment (Shillaker) [inline src/wavm/openmp.cpp:603]: This chunk of code is quite complex, but it's just been moved from higher up the file and is otherwise unchanged.

File: src/wavm/openmp.cpp (hunk 25 above)
Comment (csegarragonz) [reply src/wavm/openmp.cpp:25]: If it's cumbersome to change it's not a big deal, but with the changes we made to loggers in faasm/faabric#56, you could do something like:
```cpp
auto logger = faabric::util::getLogger(fmt::format("OMP-{}", ctx.threadNumber));
logger->debug(str);
```

**Review by csegarragonz — APPROVED** [review]
LGTM, just mentioned a very minor thing.
